Fallout 4 Tonemapping and more expansions of MagicHDR

MaxG2D opened the pull request:
All credit goes to kingeric1992 at ENB Forums for deciphering the formula (http://enbseries.enbdev.com/forum/viewtopic.php?f=7&t=4695).

Also, added a naive fix to a Issue #3 , reusing the OpenGL route, otherwise, it was throwing the same error in Fo4 on 3440x1440 resolution.

MaxG2D committed:
Added Fallout 4 tonemapping to MagicHDR.FX

MaxG2D committed:
Update Tonemap.fxh with Fallout 4 tonemapping

MaxG2D commented:
**Before:**
![Fallout4 2023-02-19 00-47-40](https://user-images.githubusercontent.com/88550439/219904615-0a738f72-6e9c-4dfa-9b6b-c22ab31ad294.jpg)
**After:**
![Fallout4 2023-02-19 00-47-38](https://user-images.githubusercontent.com/88550439/219904612-8c7e64b3-7f56-4a82-9b67-4f5ee96283f7.jpg)

MaxG2D committed:
Fixed whitepoint being applied very incorrectly

I'm a moron, it now works properly and gives proper HDR values\Bloom tones

MaxG2D commented:
**Before:**
![Fallout4 2023-02-19 02-35-15](https://user-images.githubusercontent.com/88550439/219907423-6f797630-c294-4ff3-875c-e147ac6ab51f.jpg)
**After:**
![Fallout4 2023-02-19 02-35-13](https://user-images.githubusercontent.com/88550439/219907430-af5b031d-4078-4748-878f-da50b6c8a8d9.jpg)

luluco250 commented:
Firstly, thanks for the PR.

Full disclosure: I've not at all been active in ReShade shader development for well over an year now, so I apologize in advance if I come off a little rusty or dismissive, that is not my intention at all.

Looking over the tonemap algorithm, it seems to me that it is simply the Uncharted 2 tonemapper with different constants. Normally in cases like this where two things are so similar, I'd prefer to merge both with some sort of macro or function that can have the constants as parameters.

Also the reason for the UC2 tonemapper port to lack a W parameter is simply because it is canonically set to `1.0`, which makes it redundant in the algorithm (obviously not if you *do* change it).

Regarding the random color shifts/precision loss, that is the same case with the UC2 tonemapper. Maybe it's due to there being way too many variables involved (it *is* a rather convoluted algorithm, just compare it to ACES).

Lastly, about the fix for #3, if I understand correctly all it does is set the adaptation texture to always have a single mip level outside of OpenGL. That'd break the whole point of using mipmaps though, no? The reason for them is to get an average of the entire texture, which is possible through the last mip level of a texture. When working with OpenGL it seems it tends to not like non power of two (NPOT) textures when using mipmaps, but I thought maybe it was not an issue with DirectX since it did not break *for me*.

Maybe always using the OpenGL path would work? It'd make the last texture 1:1 rather than follow the aspect ratio of the game though.

I can resolve some nitpicks with further commits while pulling the PR though so don't worry about that.

MaxG2D commented:
Thanks for the detailed reply!

The color loss was mostly me being stupid and doing things wrong, with my last commit, it's very close to what ENB would output, AKA, real HDR bloom, so I'm pretty happy with the result.

It is indeed Uncharted 2 function, with different constant values and added White point parameter, that's how F4 tonemapping works apparently. While it totally could be made by promoting the static floats to UI configuration, I felt like it would be more user friendly if I would supply it as a preset. Not everybody will dig enough to get the knowledge of how F4 tonemapper works.

Regarding the compiling problem, you're right it would be better to simply set it to 1:1 pixels. Not really sure what's causing the problem for Reshade, this error is weird, cause it works just fine in some games, but doesn't in others.

Anyways, I just thought I'd share my stuff and some people might enjoy it. BTW, thank you very much for you work, it's some awesome stuff and it allows to create some very nice results!

MaxG2D committed:
Major update to Tonemap.fxh

MaxG2D committed:
Major update to MagicHDR.fx

MaxG2D commented:
@luluco250 Hey, I've worked on it a bit more, so here's a bunch of key notes:

- 6 new TMOs in total, such as: new Reinhard and Lottes version that allow for more control (however, for now, it just "static const", needs better implementation to be editable in UI),
- Added GT Uchimura TMO,
- Updated Fallout 4 TMO to make sure it does not artifact on extreme high white values with "out of range" errors,
- Added whitepoint to Uncharted2 TMO, everywhere I looked, it had whitepoint implemented with "11.2" value, so assuming that it needs to be "1.0" seems incorrect,
- Added experimental AMD Luma Preserving Mapper that is full written by ChatGPT AI, so I can not guarantee that it is a proper implementation, but from my testing, it seems ROUGHLY correct.
- Probably some other stuff that I forgot, mostly clamp and saturation to ensure TMOs stay in range

MaxG2D committed:
Added bloom edge fix to MagicHDR.fx

It fixes the problem of final bloom texture edges being overexposed, especially visible when using high blur radius.

MaxG2D committed:
Change default value of edge clamping from "20.0" to "10.0"

It needs to be a bit more aggressive.

MaxG2D committed:
One more update to bloom edges clamping

Crossed fingers it's the last one, sorry for the spam

MaxG2D committed:
Removed clamp from Lottes tonemapping, it was unnecessary

MaxG2D committed:
Added bloom contrast function

Usefully for getting rid of the "haze" look, without changing the way the bloom texture looks, it is done in preprocess stage before the bloom is applied.

MaxG2D committed:
Update to bloom contrast code

MaxG2D committed:
Added better method of clamping bloom texture edges

By applying "border" Address UV for sampler. 
Keeping old clamping method as well, it still might be of some use if somebody wants some vignetting to the bloom

MaxG2D committed:
Changed adaptation to 1x1 pixel regardles of API

MaxG2D committed:
Added one more TMO, iCAM06m

MaxG2D committed:
Removed legact edge clamp

It's no longer needed once I added border UV to sampling in one of the previous commits

MaxG2D commented:
Hey @luluco250, I think I've pushed a last update, don't see much to add anymore to this wonderful shader of yours. Out of notable new changes, I've added border UV to sampler, so that gets rid of bloom going crazy at the screen edges, added optional contrast function to bloom texture, and as you suggested, 1x1 pixel adaptation from OpenGL implementation is now used in every API to avoid #3 issue. Feel free to merge, or not, no pressure, and thanks for making all the awesome shaders for ReShade throughout the years :)

MaxG2D committed:
More Changes

- Changes the way saturation works, no longer increasing brightness

MaxG2D committed:
More Changes

- Minor cleanup in most of the TMOs,
- More accurate Fallout 4 TMO,
- Added Linear "tonemapping", mostly for testing

MaxG2D committed:
More Changes

- Added setting for bloom blending, so now it can be either overlay (original) or additive,
- More precise conversion between linear/sRGB, default Reshade implementation is quite in precise, and it's pretty important to get that right for inverse tonemapping

MaxG2D committed:
Fix for Reshade 5.9.0+

MaxG2D committed:
One more update to saturation function

Works even better now in terms of preserving the scale of the input, and adjusting solely the saturation of the input

MaxG2D committed:
HDR display support (Tonemap.fxh)

- Removed completely buggy AMD LPM,
- Adjusted all tonemappers to remove all clamps\saturate, to be more compatible with HDR displays (it is now controlled in MagicHDR.fx),
- Added "none" tonemapping to act as a bypass, definitely useful in HDR displays,
- Added my awful attempt at porting Dice's Frostbite TMO,
- Ported hue preserving tonemapping from: https://www.shadertoy.com/view/fsXcz4

MaxG2D committed:
HDR display support (MagicHDR.fx)

- Added HDR display support, it is now controlled by setting up max nits value for both inverse and apply tonemapping,
- Added alternative bloom brightness setting ("max" operator), it helps to apply some contrast without color shifting
- Removed log from additive bloom amount output, seems to work better for additive blending,
- Changed some default\max\min values of UI settings to my own personal preference when working on HDR display.

MaxG2D commented:
Just a heads up @luluco250, I've recently got myself a HDR display, and I've did some more edits to make MagicHDR compatible with HDR input\output. Feel free it use it however you want, it works pretty well for my use case, perhaps it will be also useful to somebody else :)

luluco250 commented:
Thanks, I'll keep it in mind.

Only reason I haven't merged is because I'm not currently working with ReShade so I don't have time to review/integrate it, but thanks a lot!

MaxG2D committed:
Small update to F4 tonemapping

MaxG2D committed:
One more fix for HDR display

Removed NormalDistribution function from bloom texture, it was obviously clamping the bloom texture to 80 nits, and we don't want that in HDR displays.

MaxG2D committed:
Update

- New contrast thresholding (before: smoothstep, after: luminance) that's a bit more tricky to set up but gives better results with no color shifting,
- Brought back normal distribution, but removed line that was dividing the output,
- Removed exponent for BloomBrightness, but only in additive bloom mode. It helps to make it more linear, which is what we want in true HDR.

MaxG2D committed:
Better customization\UI options, cleanup

Shader is now very customizable and cleaner as a result
